system/Database/BaseBuilder.php

```diff
@@ -1785,21 +1795,13 @@ public function insertBatch(?array $set = null, ?bool $escape = null, int $batch
     }
 
     /**
-     * Generates a platform-specific insert string from the supplied data.
-     */
-    protected function _insertBatch(string $table, array $keys, array $values): string
-    {
-        return 'INSERT ' . $this->compileIgnore('insert') . 'INTO ' . $table . ' (' . implode(', ', $keys) . ') VALUES ' . implode(', ', $values);
-    }
-
-    /**
-     * Allows key/value pairs to be set for batch inserts
+     * Allows key/value pairs to be set for batch inserts/upserts
      *
-     * @param mixed $key
+     * @param array|object $key
      *
      * @return $this|null
      */
-    public function setInsertBatch($key, string $value = '', ?bool $escape = null)
+    public function setBatch($key, string $value = '', ?bool $escape = null)
```

Comment:
Do you really need `$value`? For what?

Reply:
The setInsertBatch method accepts key value pairs if $key is a string.
```php
if (! is_array($key)) {
            $key = [$key => $value];
        }
```
 Im really not sure why being a batch operation though. Im looking at a way to consolidate all the setBatch methods and this presents some issues. The setUpdateBatch() method accepts $index as a second parameter but it doesnt do much of anything with it.

Reply:
See #6315
The current batch APIs are broken or not consistent.
So I don't think it is necessary to match what is broken.

I don't know why one key/value string pair is needed for batch operation, and even if you need to set one record,
you can set as `[[key => value]]`.

> The setUpdateBatch() method accepts $index as a second parameter but it doesnt do much of anything with it.

It may indicate to consolidate all the setBatch methods is not reasonable.

I think the following is enough:
```php
setUpsertBatch($set, $escape)
```
And make the `$value` of other batch methods deprecated, and will remove in v5.
What do you think?

Reply:
@kenjis 
I spent a lot of time looking at the code this past weekend and thinking how this should all work out. There are definitely some inconsistencies. I'd like to refactor everything.. perhaps after this PR is committed. Basically I think we should start from the ground up with the `setBatch()` method and make it work for `insertBatch()`, `upsertBatch()`, `updateBatch()`, and a possible new one `deleteBatch()`. Then the old `setUpdateBatch()`, `setInsertBatch()`, etc we can keep functioning as is and make them depreciated. 

One thing I noticed is that `setUpdateBatch()` and `updateBatch()` seems to only allow for one field to be used as $index. So if you have a dataset with composite primary keys then you couldn't update them in a batch. The update() method allows using a where() clause and so It would work for multiple keys. 

The `updateBatch()` SQL for MySQL is pretty ugly. I have a different method for writing SQL for batch updates. I'll run some performance tests to see how it performs against the existing methods.

```SQL
UPDATE table_name
INNER JOIN (
SELECT 1 primary1, 1 primary2, 10 column3, 20 column4, 30 column5 UNION ALL
SELECT 1 primary1, 2 primary2, 40 column3, 50 column4, 60 column5
) updates
  ON table_name.primary1= updates.primary1 # these are your $constraints
  AND table_name.primary2= updates.primary2
SET
  table_name.column3 = updates.column3,
  table_name.column4= updates.column4,
  table_name.column5= updates.column5; 
```
This same query design could be used for `deleteBatch()` where you have multiple keys you are working with.

And then to be consistent with upserts I think that the `onConstraint()` method may be used to set the fields to update on. 

for the above query a statement might look like:

```php
$builder->setBatch($data)->onConstraint('primary1, primary2')->updateBatch();
// or 
$builder->setBatch($data, 'primary1, primary2', 50)->updateBatch();
// or 
$builder->updateBatch($data, 'primary1, primary2', 50);
// or 
$builder->setBatch($data)->onConstraint('primary1, primary2')->updateBatch(null,null,50);
```
Basically I was thinking that the `setBatch()` method signature should should align with the `xBatch()` methods. The set method basically is just used when you want to set the data but not execute the query yet. Perhaps you use the `setBatch()` method in a loop collecting rows.

So: 
```php
$builder->setBatch($set, $constraint = null, $batchsize = 100, $escape = true);
// and
$builder->updateBatch($set = null, $constraint = null, $batchsize = 100, $escape);
// and
$builder->insertBatch($set = null, $batchsize = 100, $escape);
// and
$builder->upsertBatch($set = null, $constraint = null, $batchsize = 100, $escape);
// and
$builder->deleteBatch($set = null, $constraint = null, $batchsize = 100);
// OR reorder them so $constraint is last being its not used for inserts
$builder->setBatch($set, $batchsize = 100, $escape = true, $constraint = null);
// OR remove paramater and force to use other methods to set them.
$builder->setBatch($set, $escape = true)->onConstraint($constraint)->setBatchSize($batchSize)->updateBatch();
```
This would work for insert,update,upsert,delete Batch. When using with `insertBatch()` the constraint parameter would be unused.

Alternatively we could drop the $constraint parameter for force you to use `onConstraint()` method.

Similarly with $batchsize. The reason to include the `setBatch()` is keep the last method clean: `->updateBatch()`;

Then I would like to commit all the $set data to $QBSet. This can be done in methods such as `insertBatch()`, etc. This way the `executeBatch()` method can be called with out a $set parameter and there doesn't need to be dual logic for $set and $QBSet. The way it currently is if $set data is used instead of $QBSet then the data isn't cleaned, escaped, etc. Also I like to see if $binds could be used to the batch data as well... currently it is not. 

As it stands, the `setBatch()` method is pretty much just a direct copy of what the `setInsertBatch()` method used to be.

So in conclusion, for now we can design this PR around the `setBatch()` method, keeping in mind for future use with the other DML statements. If we commit this now, we still expect to revise and refactor before final release. To say exactly what parameters we should use now may not be the best time. I think doing some refactoring of the code and then having a look at things then when they are all aligned together might make for a better conversation on what exactly the public API should look like.

Reply:
I changed to `setBatch($key, ?bool $escape = null)`. We still can come back and take another look later.

Reply:
> Also I like to see if $binds could be used to the batch data as well... currently it is not.

We do not use $binds with batch operations intentionally. It significantly degrades performance.
See https://github.com/codeigniter4/CodeIgniter4/pull/5202#issuecomment-944037647